core/src/impl/Kokkos_Half_MathematicalFunctions.hpp

```diff
@@ -155,10 +157,59 @@ KOKKOS_IMPL_MATH_HALF_FUNC_WRAPPER(KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF, nextaf
 KOKKOS_IMPL_MATH_HALF_FUNC_WRAPPER(KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF, copysign)
 // Classification and comparison functions
 // fpclassify
-KOKKOS_IMPL_MATH_HALF_FUNC_WRAPPER(KOKKOS_IMPL_MATH_UNARY_PREDICATE_HALF, isfinite)
-KOKKOS_IMPL_MATH_HALF_FUNC_WRAPPER(KOKKOS_IMPL_MATH_UNARY_PREDICATE_HALF, isinf)
-#if !defined(KOKKOS_ENABLE_SYCL) && !defined(KOKKOS_ENABLE_HIP) // FIXME_SYCL, FIXME_HIP
-KOKKOS_IMPL_MATH_HALF_FUNC_WRAPPER(KOKKOS_IMPL_MATH_UNARY_PREDICATE_HALF, isnan)
+
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+KOKKOS_INLINE_FUNCTION bool isinf(Kokkos::Experimental::half_t x) {
+   using bit_type = Kokkos::Experimental::half_t::bit_comparison_type;
+   bit_type bit_pattern_x   = bit_cast<bit_type>(static_cast<Kokkos::Experimental::half_t::impl_type>(x));
+   bit_type bit_pattern_inf = Kokkos::Experimental::infinity_v<Kokkos::Experimental::half_t>;
+   return (bit_pattern_x.value == bit_pattern_inf.value);
+ }
+#endif
+
+#if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
+KOKKOS_INLINE_FUNCTION bool isinf(Kokkos::Experimental::bhalf_t x) {
+   using bit_type = Kokkos::Experimental::half_t::bit_comparison_type;
+   bit_type bit_pattern_x   = bit_cast<bit_type>(static_cast<Kokkos::Experimental::bhalf_t::impl_type>(x));
+   bit_type bit_pattern_inf = Kokkos::Experimental::infinity_v<Kokkos::Experimental::bhalf_t>;
+   return (bit_pattern_x.value == bit_pattern_inf.value);
+ }
+#endif
+
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+KOKKOS_INLINE_FUNCTION bool isnan(Kokkos::Experimental::half_t x) {
+   using bit_type = Kokkos::Experimental::half_t::bit_comparison_type;
+   bit_type bit_pattern_x         = bit_cast<bit_type>(static_cast<Kokkos::Experimental::half_t::impl_type>(x));
+   bit_type bit_pattern_quiet_nan = Kokkos::Experimental::quiet_NaN_v<Kokkos::Experimental::half_t>;
+   if (bit_pattern_x.value == bit_pattern_quiet_nan.value)
+     return true;
+   bit_type bit_pattern_signaling_nan = Kokkos::Experimental::signaling_NaN_v<Kokkos::Experimental::half_t>;
+   return (bit_pattern_x.value == bit_pattern_signaling_nan.value);
+ }
+#endif
+
+#if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
+KOKKOS_INLINE_FUNCTION bool isnan(Kokkos::Experimental::bhalf_t x) {
+  using bit_type = Kokkos::Experimental::half_t::bit_comparison_type;
+  bit_type bit_pattern_x         = bit_cast<bit_type>(static_cast<Kokkos::Experimental::bhalf_t::impl_type>(x));
+  bit_type bit_pattern_quiet_nan = Kokkos::Experimental::quiet_NaN_v<Kokkos::Experimental::bhalf_t>;
+  if (bit_pattern_x.value == bit_pattern_quiet_nan.value)
+    return true;
+  bit_type bit_pattern_signaling_nan = Kokkos::Experimental::signaling_NaN_v<Kokkos::Experimental::bhalf_t>;
+  return (bit_pattern_x.value == bit_pattern_signaling_nan.value);
```

Comment:
We probably need to make a couple more values `NaN`, see https://github.com/intel/llvm/blob/ebd1d1fd296f07694f8dfe53bc7cd9b9906be5c5/sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp#L33-L39.